src/b64.ml

```diff
@@ -16,77 +19,143 @@
+
+let unsafe_set_be_uint16 =
+  if Sys.big_endian
+  then fun t off v -> unsafe_set_uint16 t off v
+  else fun t off v -> unsafe_set_uint16 t off (swap16 v)
+
+exception Out_of_bound
+
+let get_uint8 t off =
+  if off < 0 || off >= String.length t then raise Out_of_bound ;
+  unsafe_get_uint8 t off
+
+let padding = int_of_char '='
+
+let encode pad { emap; _ } input =
+  let n = String.length input in
+  let n' = n // 3 * 4 in
+  let res = Bytes.create n' in
+
+  let emap i = Array.unsafe_get emap i in
+
+  let emit b1 b2 b3 i =
+    unsafe_set_be_uint16 res i ((emap (b1 lsr 2 land 0x3f) lsl 8) lor (emap ((b1 lsl 4) lor (b2 lsr 4) land 0x3f))) ;
+    unsafe_set_be_uint16 res (i + 2) ((emap ((b2 lsl 2) lor (b3 lsr 6) land 0x3f) lsl 8) lor (emap (b3 land 0x3f))) in
+
+  let rec enc j i =
+    if i = n then ()
+    else if i = n - 1 then emit (unsafe_get_uint8 input i) 0 0 j
+    else if i = n - 2 then emit (unsafe_get_uint8 input i) (unsafe_get_uint8 input (i + 1)) 0 j
+    else
+    (emit
+       (unsafe_get_uint8 input i)
+       (unsafe_get_uint8 input (i + 1))
+       (unsafe_get_uint8 input (i + 2))
+       j ;
+     enc (j + 4) (i + 3)) in
+
+  let rec fix = function
+  | 0 -> ()
+  | i -> unsafe_set_uint8 res (n' - i) padding ; fix (i - 1) in
+
+  enc 0 0 ;
+
+  if pad
+  then begin fix ((3 - n mod 3) mod 3) ; Bytes.unsafe_to_string res end
+  else Bytes.sub_string res 0 (n / 3 * 4)
+
+let encode ?(pad = true) ?(alphabet = default_alphabet) input = encode pad alphabet input
+
+let decode_result { dmap; _ } input =
+  let n = String.length input in
+  let n' = (n / 4) * 3 in
+  let res = Bytes.create n' in
+
+  let emit a b c d i =
+    let x = (a lsl 18) lor (b lsl 12) lor (c lsl 6) lor d in
+    unsafe_set_be_uint16 res i (x lsr 8) ;
+    unsafe_set_uint8 res (i + 2) (x land 0xff) in
+
+  let dmap i =
+    let x = Array.unsafe_get dmap i in
+    if x = none then raise Not_found else x in
+
+  let rec dec j i =
+    if i = n then Some 0
+    else begin
+      let a = dmap (get_uint8 input i) in
+      let b = dmap (get_uint8 input (i + 1)) in
+      let (d, pad) =
+        let x = get_uint8 input (i + 3) in
+        try (dmap x, 0) with Not_found when x = padding -> (0, 1) in
+      let (c, pad) =
+        let x = get_uint8 input (i + 2) in
+        try (dmap x, pad) with Not_found when x = padding && pad = 1 -> (0, 2) in
+
+      emit a b c d j ;
+      if pad = 0
+      then dec (j + 3) (i + 4)
+      else if i + 4 <> n then None
+      else Some pad end in
+
+  match dec 0 0 with
+  | Some pad -> Ok (Bytes.sub_string res 0 (n' - pad))
+  | None -> Error `Wrong_padding
+  | exception Out_of_bound ->
+      (* appear when [get_uint8] wants to access to an invalid area *)
+      Error `Wrong_padding
+  | exception Not_found ->
+      (* appear when [dmap] not found associated character.
+         an other branch is when we got '=' (so [dmap] returns [Not_found]) and the last
+         character is not an '='. *)
+      Error `Malformed
```

Comment:
Can a caller meaningfully do anything different if the input is Malformed _vs_ wrong padding?  It might be better to return a `Error (Msg "Wrong padding")` and `Error (Msg "Malformed")` so that it is compliant with other Mirage rresult errors